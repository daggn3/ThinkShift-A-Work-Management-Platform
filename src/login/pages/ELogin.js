import React, {Component} from 'react';
import { withStyles  } from "@material-ui/core/styles";
import {Avatar, Button, CssBaseline, TextField,Grid,Box,Typography,Container} from '@material-ui/core';
import SupervisedUserCircleIcon from '@material-ui/icons/SupervisedUserCircle';
import Copyright from "../components/copyright"
import Snackbar from '@material-ui/core/Snackbar';
import axios from "axios"
import { Redirect } from "react-router"
import MuiAlert from '@material-ui/lab/Alert';



//Material UI theme//
const styles = theme => ({
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.primary.main,
  },
  form: {
    width: '100%', 
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
});


class ELogin extends Component {
  
  constructor(props) {
    super()


    //states needed for logins
    this.state = {
      open : false,
      email:"",
      password:"",
      fireRedirect: false
    }


    //our binders for our functions
    this.changeEmail = this.changeEmail.bind(this)
    this.changePassword = this.changePassword.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.handleClick = this.handleClick.bind(this)
  }

  //changes email state to current value
  changeEmail(event) {
    this.setState({
      email:event.target.value
    })
  }

  //changes password to current value
  changePassword(event) {
    this.setState({
      password:event.target.value
    })
  }

  //changes open state for our snackbar
  handleClick = () =>{
    this.setState({open: true})
  }

  //handles a clickaway close for our snackbar
  handleClose = (reason) => {
    if(reason ==="clickaway") {
      return
    }
    this.setState({open: false})
  }

  //submit and send to database to recieve token and response
  onSubmit(event){
    event.preventDefault()

    const signin = {
      email: this.state.email,
      password: this.state.password
    }
    
    //Posts info to database, if 200 response, redirects to home
    axios.post("http://ec2-34-245-24-4.eu-west-1.compute.amazonaws.com:5000/login/employee", signin, {withCredentials:true})
    .then(response => { 
      this.setState({
      email:"",
      password:"",
      fireRedirect: true,
      })
      
      //sets our props login state to "employee"
      this.props.Logintype("employee")

      //take in employee info for use in our employee timetable component
      window.sessionStorage.setItem("name", response.data.name)
      window.sessionStorage.setItem("id",response.data._id )
      
      
    }).catch(e => {
      this.setState({
        open:true
      })
      
    })
    

  }


  render() {

    const { classes } = this.props;
    const { fireRedirect } = this.state
    

  return (



    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <SupervisedUserCircleIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          ThinkShift Employee Sign in
        </Typography>
        <form onSubmit={this.onSubmit} className={classes.form}>
          <TextField
            onChange={this.changeEmail}
            value={this.state.email}
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="email"
            label="Email Address"
            name="email"
            autoComplete="email"
            autoFocus
          />
          <TextField
            onChange = {this.changePassword}
            value={this.state.password}
            variant="outlined"
            margin="normal"
            required
            fullWidth
            name="password"
            label="Password"
            type="password"
            id="password"
            autoComplete="current-password"
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            Sign In
          </Button>


          
          <Button
            variant="link"
            href="/"
            fullWidth
            variant="contained"
            color= "secondary"
            className={classes.submit}
          >
            Manager Login
          </Button>
          
          

          <Snackbar
            anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'center',
          }}

           open={this.state.open} autoHideDuration={6000} onClose={this.handleClose}>

          <MuiAlert elevation={6} variant="filled" onClose={this.handleClose} severity="error">
            Login Failed! Please Try Again.
          </MuiAlert>

          </Snackbar>

          <Grid container>
            <Grid item xs>


            </Grid>
            <Grid item>
            </Grid>
            <Grid item>
            </Grid>
          </Grid>
        </form>

        {/* redirects to home */}
        {fireRedirect && (
          <Redirect to={"/time"}/>
        )}
      </div>
      <Box mt={8}>
      <Copyright />
      </Box>
    </Container>
  );
}}

export default withStyles(styles, { withTheme: true })(ELogin);