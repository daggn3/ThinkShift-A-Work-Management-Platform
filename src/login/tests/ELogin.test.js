import ReactDOM from "react-dom"
import ELogin from "../pages/ELogin"
import {render, cleanup, shallow, mount, act, fireEvent, getByText} from "@testing-library/react"
import renderer from "react-test-renderer"

afterEach(cleanup)

it("renders without crashing", () => {
  const elog = document.createElement("elog")
  ReactDOM.render(<ELogin />, elog)
})

//quick tests to make sure Text fields render


it("renders correctly", () => {
    const {getByTestId} = render(<ELogin />)
    expect(getByTestId("email")).toHaveTextContent("Email Address")
    expect(getByTestId("password")).toHaveTextContent("Password")
})

describe("Login", () => {
  describe("with valid inputs", () => {
    it('calls the onSubmit function', async () => {
      const mockOnSubmit = jest.fn()
      const {getByLabelText, getByText} = render(<ELogin onSubmit={mockOnSubmit}/>)

      await act(async () => {
        
        fireEvent.change(getByLabelText("Email"), {target: {value: "test@gmail.com"}})
        fireEvent.change(getByLabelText("Password"), {target: {value: "1234"}})
      })

      await act(async () => {
        fireEvent.click(getByText("button"))
      })

      expect(mockOnSubmit).toHaveBeenCalled()
    })
  })
})


it("matches snapshot", () => {
    const tree = renderer.create(<ELogin/>).toJSON()
    expect(tree).toMatchSnapshot()
})