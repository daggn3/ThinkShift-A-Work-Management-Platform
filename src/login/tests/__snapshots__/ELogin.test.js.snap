// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    className="ELogin-paper-1"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle ELogin-avatar-2 MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.99 2c-5.52 0-10 4.48-10 10s4.48 10 10 10 10-4.48 10-10-4.48-10-10-10zm3.61 6.34c1.07 0 1.93.86 1.93 1.93 0 1.07-.86 1.93-1.93 1.93-1.07 0-1.93-.86-1.93-1.93-.01-1.07.86-1.93 1.93-1.93zm-6-1.58c1.3 0 2.36 1.06 2.36 2.36 0 1.3-1.06 2.36-2.36 2.36s-2.36-1.06-2.36-2.36c0-1.31 1.05-2.36 2.36-2.36zm0 9.13v3.75c-2.4-.75-4.3-2.6-5.14-4.96 1.05-1.12 3.67-1.69 5.14-1.69.53 0 1.2.08 1.9.22-1.64.87-1.9 2.02-1.9 2.68zM11.99 20c-.27 0-.53-.01-.79-.04v-4.07c0-1.42 2.94-2.13 4.4-2.13 1.07 0 2.92.39 3.84 1.15-1.17 2.97-4.06 5.09-7.45 5.09z"
        />
      </svg>
    </div>
    <h1
      className="MuiTypography-root MuiTypography-h5"
    >
      ThinkShift Employee Sign in
    </h1>
    <form
      className="ELogin-form-3"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="email"
          id="email-label"
        >
          Email Address
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="email"
            autoFocus={true}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="email"
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Email Address
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="current-password"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Password
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained ELogin-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
        data-testid="button"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Sign In
        </span>
      </button>
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-contained ELogin-submit-4 MuiButton-containedSecondary MuiButton-fullWidth"
        href="/"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <span
          className="MuiButton-label"
        >
          Manager Login
        </span>
      </a>
      <div
        className="MuiGrid-root MuiGrid-container"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
        />
        <div
          className="MuiGrid-root MuiGrid-item"
        />
        <div
          className="MuiGrid-root MuiGrid-item"
        />
      </div>
    </form>
  </div>
  <div
    className="MuiBox-root MuiBox-root-12"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
    >
      Copyright © 
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
        href="http://ec2-34-245-24-4.eu-west-1.compute.amazonaws.com:3000"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        ThinkShift
      </a>
       
      2021
      .
    </p>
  </div>
</main>
`;
